vsphere_plugin_common/__init__.py

```diff
@@ -766,43 +850,133 @@ def get_dvswitches(self):
+                        'No valid vswitches found. '
+                        'Every physical host in the datacenter must have the '
+                        'same named vswitches available when not using '
+                        'distributed vswitches.'
+                    )
+                else:
+                    raise cfy_exc.NonRecoverableError(
+                        '{vswitch} was not a valid vswitch name. The valid '
+                        'vswitches are: {vswitches}'.format(
+                            vswitch=vswitch_name,
+                            vswitches=', '.join(vswitches),
+                        )
+                    )
+
+            if self.port_group_is_on_any_hosts(port_group_name):
                 raise cfy_exc.NonRecoverableError(
-                    'No valid vswitches found. '
-                    'Every physical host in the datacenter must have the '
-                    'same named vswitches available when not using '
-                    'distributed vswitches.'
+                    'Port group {name} already exists on vSphere.'.format(
+                        name=port_group_name,
+                    )
                 )
+
+        if runtime_properties['status'] in ('preparing', 'creating'):
+            runtime_properties['status'] = 'creating'
+            if 'created_on' not in runtime_properties.keys():
+                runtime_properties['created_on'] = []
+
+            hosts = [
+                host for host in self.get_host_list()
+                if host.name not in runtime_properties['created_on']
+            ]
+
+            for host in hosts:
+                network_system = host.configManager.networkSystem
+                specification = vim.host.PortGroup.Specification()
+                specification.name = port_group_name
+                specification.vlanId = vlan_id
+                specification.vswitchName = vswitch_name
+                vswitch = network_system.networkConfig.vswitch[0]
+                specification.policy = vswitch.spec.policy
+                ctx.logger.debug(
+                    'Adding port group {group_name} to vSwitch '
+                    '{vswitch_name} on host {host_name}'.format(
+                        group_name=port_group_name,
+                        vswitch_name=vswitch_name,
+                        host_name=host.name,
+                    )
+                )
+                try:
+                    network_system.AddPortGroup(specification)
+                except vim.fault.AlreadyExists:
+                    # We tried to create it on a previous pass, but didn't see
+                    # any confirmation (e.g. due to a problem communicating
+                    # with the vCenter)
+                    # However, we shouldn't have reached this point if it
+                    # existed before we tried to create it anywhere, so it
+                    # should be safe to proceed.
+                    pass
```

Comment:
Yeah, this is something I feel like we'll be able to make cleaner once we can work on idempotency more and get more used to its quirks and can come up with a consistent style (can't be consistent when there's only one of it because there's nothing for it to be consistent with!).